articles/regular-expressions-in-python/index.md

```diff
@@ -0,0 +1,269 @@
+
+Phew.
+
+![Meme](/engineering-education/regular-expressions-in-python/meme.png)
+
+*Source: [Twitter](https://pbs.twimg.com/media/DLYTiNqW4AAvcvg.jpg:large)*
+
+Writing expressions in a certain way? What does that mean?
+
+### Syntax of Regular Expressions in Python
+
+To lay the foundation, most characters and letters are matched to themselves. That means "p" is matched to the character "p". However, there are exceptions to this, and these characters are called metacharacters. Here are the list of metacharacters:
+
+. ^ $ * + ? { } [ ] \ | ( )
+
+Before we get into the syntax, let's import Python's RegEx library, re.
+
+```py
+import re
+```
+
+Python's RegEx module offers a set of functions that search for a string for a match
+
+- findall	- Return a list of all non-overlapping matches in the string
+- search	- Looks for a match to the pattern and returns a match object, or None if no match was found.
+- split	    - Splits the source string by the occurrences of the pattern and returns a list containing the resulting substrings
+- sub	    - Replaces one or many matches in the source string
+
+The metacharacteres mentioned above can be interpreted as follows:
+
+1. Square brackers denote a character class, a set of characters that are to be matched. "[abef]" maps to any one of "a", "b", "e", "f". If the comparison is for a character within a range, the starting and the ending characters must be specified within the square brackets, separated by a "-". "[a-z]" maps to any character in lowercase and "[A-Z]" maps to any character in uppercase.
+
+    ```py
+    test_string = "I d0n't want any num6er 1n th1s 5tr1ng"
+    regex = re.findall("[0-9]",test_string)
+    if regex:
+        print("There is(are) number(s) in the string")
+    else:
+        print("There are no numbers in this string")
+
+    # Output - There is(are) number(s) in the string
+    ```
+
+2. Another rule that metacharacters enforce is that they do not have any effect in a character class unless it starts with a metacharacter. A complement of a character class is denoted by the metacharacter "^". So, "[^abc]" will match with any character except "a", "b" and "c" whereas "[abc^]" will match with characters "a", "b", "c" and "^" even though "^" is a metacharacter. Note that the complement happens only within a character class.
+
+    ```py
+    test_string = "This string has no e in it"
+    regex = re.search("[^e]",test_string)
+    if regex:
+        print("There is an e in the string")
+    else:
+        print("There is no e in this string")
+
```

Comment:
I find this example problematic because the print statements imply that we're searching for "e" in the test string where as we are searching for any character except "e"